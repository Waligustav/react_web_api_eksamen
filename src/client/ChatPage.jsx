import React, { useState } from "react";
import { ChatView } from "./Views/ChatView";

export function ChatPage() {
  const [username, setUsername] = useState();
  if (!username) {
    return <ChatLoginPage onLogin={(username) => setUsername(username)} />;
  }

  return <ChatView username={username} />;
}

function ChatLoginPage({ onLogin }) {
  const [username, setUsername] = useState("");
  function handleSubmit(e) {
    e.preventDefault();
    onLogin(username);
  }
  return (
    <div className="mainContainer">
      <h1>Enter a chat username</h1>
      <form onSubmit={handleSubmit}>
        <input
          type="text"
          value={username}
          onChange={(e) => setUsername(e.target.value)}
        />
        <button>Login</button>
      </form>
    </div>
  );
}
