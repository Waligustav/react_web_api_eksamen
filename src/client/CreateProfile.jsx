import React, { useState } from "react";
import { InputField } from "./Components/InputField";
import { useHistory } from "react-router";
import { useSubmit } from "./Lib/useSubmit";

export function CreateProfile({ userApi }) {
  const [firstName, setFirstName] = useState("");
  const [lastName, setLastName] = useState("");
  const [email, setEmail] = useState("");
  const history = useHistory();

  const { handleSubmit, submitting, error } = useSubmit(
    async () => {
      await userApi.createUser({ firstName, lastName, email });
    },
    () => history.push("/")
  );

  return (
    <form onSubmit={handleSubmit} className="mainContainer">
      {submitting && <div>Please wait</div>}
      {error && <div>Error: {error.toString()}</div>}
      <h1>Create new profile</h1>
      <InputField
        type="text"
        value={firstName}
        label={"First name"}
        onChangeValue={setFirstName}
      />
      <InputField
        type="text"
        value={lastName}
        label={"Last name"}
        onChangeValue={setLastName}
      />
      <InputField
        type="email"
        value={email}
        label={"Email"}
        onChangeValue={setEmail}
      />
      <button disabled={submitting}>Submit</button>
    </form>
  );
}
