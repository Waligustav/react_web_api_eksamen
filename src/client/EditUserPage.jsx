import React, { useState } from "react";
import {useHistory, useParams} from "react-router";
import { LoadingView } from "./Views/LoadingView";
import { InputField } from "./Components/InputField";
import { useLoading } from "./Lib/useLoading";
import { ErrorView } from "./Views/ErrorView";

function EditUserForm({ user, onSubmit }) {
  const [firstName, setFirstName] = useState(user.firstName);
  const [lastName, setLastName] = useState(user.lastName);
  const [email, setEmail] = useState(user.email);

  async function submit(e) {
    onSubmit(e, { firstName, lastName, email });
  }

  return (
    <form onSubmit={submit} className="mainContainer">
      <h1>Edit an existing user ({firstName})</h1>
      <InputField
        label={"First name"}
        value={firstName}
        onChangeValue={setFirstName}
      />
      <InputField
        label={"Last name"}
        value={lastName}
        onChangeValue={setLastName}
      />
      <InputField
        label={"Email"}
        value={email}
        onChangeValue={setEmail}
        type="text"
      />
      <button>Submit</button>
    </form>
  );
}

export function EditUserPage({ userApi }) {
  const { id } = useParams();
  const history = useHistory();

  const { data: user, loading, error, reload } = useLoading(
    async () => await userApi.getUser(id),
    [id]
  );

  async function handleSubmit(e, { firstName, lastName, email }) {
    e.preventDefault();
    await userApi.updateUser(id, { firstName, lastName, email });
    history.push("/home");
  }

  if (error) {
    return <ErrorView error={error} reload={reload} />;
  }

  if (loading || !user) {
    return <LoadingView />;
  }

  return <EditUserForm user={user} onSubmit={handleSubmit} />;
}
