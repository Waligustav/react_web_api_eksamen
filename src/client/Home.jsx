import React from "react";
import { Link } from "react-router-dom";

export const Home = () => {
  return (
    <div className="mainContainer">
      <h1>Overview</h1>
      <li>
        <Link to="/profile">Current user</Link>
      </li>
      <h3>-- Profiles section --</h3>
      <li>
        <Link to="/users">Show profiles</Link>
      </li>
      <li>
        <Link to={"/create"}>Create profile</Link>
      </li>
      <h3>-- Chat section --</h3>
      <li>
        <Link to="/chat">Chat</Link>
      </li>
    </div>
  );
};
