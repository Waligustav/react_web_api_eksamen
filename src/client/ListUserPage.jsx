import React from "react";
import { LoadingView } from "./Views/LoadingView";
import { Link } from "react-router-dom";
import { useLoading } from "./Lib/useLoading";
import { ErrorView } from "./Views/ErrorView";

export function ProfileListPage({ userApi }) {
  const { data: users, error, loading, reload } = useLoading(
    async () => await userApi.listProfiles()
  );

  if (error) {
    return <ErrorView error={error} reload={reload} />;
  }

  if (loading || !users) {
    return <LoadingView />;
  }

  return (
    <>
      <div className="mainContainer">
        <div>
          <h1>List of profiles</h1>
          <h3>Click to edit a user</h3>
        </div>
        {users.map(({ id, firstName }) => (
          <li key={id}>
            <Link to={`/users/${id}/edit`}>{firstName}</Link>
          </li>
        ))}
      </div>
    </>
  );
}
