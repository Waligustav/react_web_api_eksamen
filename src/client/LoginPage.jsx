import * as React from "react";
import { fetchJSON } from "./Lib/http";

export function LoginPage({ identityProvider }) {
  const { discoveryURL, client_id } = identityProvider;

  async function handleLogin() {
    const { authorization_endpoint } = await fetchJSON(discoveryURL);
    const params = {
      client_id,
      response_type: "token",
      scope: "openid email profile",
      redirect_uri: window.location.origin + "/login/callback",
    };
    window.location.href =
      authorization_endpoint + "?" + new URLSearchParams(params);
  }

  return (
    <div className="mainContainerLogin">
      <h1>Login with Google</h1>
      <button onClick={handleLogin}>Login</button>
    </div>
  );
}
