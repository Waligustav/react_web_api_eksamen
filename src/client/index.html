<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="UTF-8" />
    <title>Hello there</title>
    <link rel="stylesheet" href="assets/css/style.scss">
  </head>
  <body>
    <div id="app"></div>
  </body>
  <script src="index.jsx"></script>
</html>
