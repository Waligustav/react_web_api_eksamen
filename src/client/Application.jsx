import React, { useState } from "react";
import { BrowserRouter, Link, Redirect } from "react-router-dom";
import { Route, Switch } from "react-router";
import { Home } from "./Home";
import { NotFound } from "./Views/NotFound";
import { LoginPage } from "./LoginPage";
import { ChatPage } from "./ChatPage";
import { fetchJSON, postJSON } from "./Lib/http";
import { ProfileListPage } from "./ListUserPage";
import { CreateProfile } from "./CreateProfile";
import { EditUserPage } from "./EditUserPage";
import { ProfilePage } from "./ProfilePage";
import { LoginCallbackPage } from "./LoginCallbackPage";

export function App() {
  const userApi = {
    listProfiles: async () =>
      await fetchJSON("http://localhost:3000/api/users"),
    getUser: async (id) =>
      await fetchJSON(`http://localhost:3000/api/users/${id}`),
    createUser: async ({ firstName, lastName, email }) => {
      return postJSON("http://localhost:3000/api/users", {
        method: "POST",
        json: { firstName, lastName, email },
      });
    },
    updateUser: async (id, { firstName, lastName, email }) =>
      postJSON(`http://localhost:3000/api/users/${id}`, {
        method: "PUT",
        json: { firstName, lastName, email },
      }),
    loadProfile: async () =>
      await fetchJSON("http://localhost:3000/api/profile", {
        headers: {
          ...(access_token ? { Authorization: `Bearer ${access_token}` } : {}),
        },
      }),
  };

  const [access_token, setAccessToken] = useState();
  const googleIdentityProvider = {
    discoveryURL:
      "https://accounts.google.com/.well-known/openid-configuration",
    client_id:
      "222539328927-e5tevsjdgegffj6s4pk6jro60hcati7f.apps.googleusercontent.com",
  };

  return (
    <BrowserRouter>
      <nav id="mainNav">
        <Link to={"/"}>
          {" "}
          <strong>{"<<< "}Homepage</strong>
        </Link>
      </nav>
      <main>
        <Switch>
          <Route exact path="/users">
            {!access_token ? (
              <Redirect to={"/"} />
            ) : (
              <ProfileListPage userApi={userApi} />
            )}
          </Route>
          <Route path="/create">
            {!access_token ? (
              <Redirect to={"/"} />
            ) : (
              <CreateProfile userApi={userApi} />
            )}
          </Route>
          <Route path="/users/:id/edit">
            <EditUserPage userApi={userApi} />
          </Route>
          <Route path="/profile">
            {!access_token ? (
              <Redirect to={"/"} />
            ) : (
              <ProfilePage userApi={userApi} />
            )}
          </Route>
          <Route exact path="/">
            {access_token ? (
              <Redirect to={"/home"} />
            ) : (
              <LoginPage identityProvider={googleIdentityProvider} />
            )}
          </Route>
          <Route path="/login/callback">
            <LoginCallbackPage
              identityProvider={googleIdentityProvider}
              onAccessToken={(access_token) => setAccessToken(access_token)}
            />
          </Route>
          <Route path="/chat">
            {!access_token ? <Redirect to={"/"} /> : <ChatPage />}
          </Route>
          <Route exact path="/home">
            {!access_token ? <Redirect to={"/"} /> : <Home />}
          </Route>
          <Route>
            <NotFound />
          </Route>
        </Switch>
      </main>
    </BrowserRouter>
  );
}
