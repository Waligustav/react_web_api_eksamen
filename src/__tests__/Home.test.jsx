import React from "react";
import ReactDOM from "react-dom";
import { act } from "react-dom/test-utils";
import { Home } from "../client/Home";
import { MemoryRouter } from "react-router";

async function testRenderer(component) {
  const container = document.createElement("div");
  await act(async () => {
    await ReactDOM.render(<MemoryRouter>{component}</MemoryRouter>, container);
  });
  return container;
}

describe("Display homepage", () => {
  it("shows homepage", async () => {
    const container = await testRenderer(<Home />);
    expect(container.innerHTML).toMatchSnapshot();
    expect(container.querySelector("h1").textContent).toEqual("Overview");
  });
});
