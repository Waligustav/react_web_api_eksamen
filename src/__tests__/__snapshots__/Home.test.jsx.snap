// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Display homepage shows homepage 1`] = `"<div class=\\"mainContainer\\"><h1>Overview</h1><li><a href=\\"/profile\\">Current user</a></li><h3>-- Profiles section --</h3><li><a href=\\"/users\\">Show profiles</a></li><li><a href=\\"/create\\">Create profile</a></li><h3>-- Chat section --</h3><li><a href=\\"/chat\\">Chat</a></li></div>"`;
